### DISCUSSION TITLE
Preserve input names

### DISCUSSION THREAD

**Discussion by hasparus** [wevm/wagmi#1977] [Ideas]
### What

Given function inputs in the API

```ts
inputs: [
  { name: 'a', internalType: 'uint64', type: 'uint64' },
  { name: 'b', internalType: 'uint64', type: 'uint64' },
],
```

it would be nice if the the type of `args` was a labeled tuple

```ts
[a: BigNumber, b: BigNumber]
```

or an object

```ts
{ a: BigNumber, b: BigNumber }
```

instead of a tuple unnamed args

```ts
[BigNumber, BigNumber]
```

so it's harder to make a mistake in the order of arguments.

### Why

It would be helpful in adopting in frontend codebases currently using TypeChain generated `ethers` Contract instances.

### How

Supporting objects of args is possible on type level during dev-time.

```ts
type AbiParametersToArgsObject<TAbiParameters extends readonly AbiParameter[]> = MergeUnion<{
  [I in keyof TAbiParameters]: { [N in TAbiParameters[I]['name'] & string]: AbiParameterToPrimitiveType<TAbiParameters[I]> }
}[Indices<TAbiParameters>]>

type _result = AbiParametersToArgsObject<typeof inputs>
   // ^? { a: bigint, b: bigint }
```

It would give the most typesafety, but it would require mapping from `inputs` and `args` object to `args` array to send the transaction.

```ts
const argsArray = inputs.map(abiParameter => argsObject[abiParameter.name])
```

[TypeScript Playground](https://www.typescriptlang.org/play?ts=4.9.5#code/JYWwDg9gTgLgBAbzgQQEbAAoEMpZAUxnygBUIMpRgZgA3fEgTzHwBoV1tcCip2BfOADMoEEHADkWdDGb4JAbgBQKgMYQAdgGd4wDWACuMLXAC8cANpK4iOBrz4AXJKwT2e3vYA2TFs4kGHgBsACxucLJ+koEaMKEScPys1rb2BP6o4R7E3r5O0cFh7JH5AYUJSUoAunBYJuraMMoqJRyYOA68WmTIUADmWgDyqABW+KowADwkaO3chMQm+AAeRBoAJiZQ+Fjrml6MbVydxBZVAHxmcACyxH34AKoawJqTCCkWAJJwenAA1vhGBAhHAZpwOjxFlVnEgLAA5H4aUGzY6QqBaL5VCwSNLyGoAMjgOkoGj60KOEIWpHIlBA1DoDDk0xRlK6mMu-CU-C+G2AqnwWmZ4PmXXOFxUSlaAH1tloDF54OYWSLFj1+kNRuMpiVgYjDMZzik4AB6Y1wAB6AH4JSpTXAALSOm2SuRwT68-mCkhwFZrTZwba7faHAwaP4aCAAdw0Z0u5gAostVF4DOt8JMAUCQSR2JndbGWq7bv1Hs9XiQ44gPhhEaCfat8BsTFgNIdLf9Abrvc4NPh6FBoSlvb7G-7YX9axgqpbnC3DpybO2SBYpyke33iFylEA)

AFAIK Labeled tuples would have to be codegenned, and they don't enforce the order, but they wouldn't require supporting two syntaxes.

**Comment by hasparus** [comment]
Alternatively, to get args as a labeled tuple, Wagmi CLI could codegen a phantom decoration like GraphQL Codegen's TypedDocumentNode.

**Comment by tmm** [comment]
Appreciate you writing this up! Definitely think this would be a moderate improvement, but it's not as robust dynamic labeled tuples since it adds runtime code and makes it hard to backtrack to dynamic labeled tuples if/when they are supported. If it doesn't seem like dynamic labeled tuples will ever be a feature in TypeScript, then it might be time to revisit this — but I'm optimistic we'll see it at some point!
